modules/post/multi/manage/shell_to_meterpreter.rb

```diff
@@ -249,24 +264,34 @@ def transmit_payload(exe, platform)
       #
       sent = 0
       aborted = false
-      cmds.each do |cmd|
-        ret = cmd_exec(cmd)
-        if !ret
-          aborted = true
-        else
-          ret.strip!
-          aborted = true if !ret.empty? && ret !~ /The process tried to write to a nonexistent pipe./
-        end
-        if aborted
-          print_error('Error: Unable to execute the following command: ' + cmd.inspect)
-          print_error('Output: ' + ret.inspect) if ret && !ret.empty?
-          break
-        end
 
-        sent += cmd.length
+      if cmds.length >= 2
+        cmds.take(cmds.length - 1).each do |cmd|
+
+          # These commands need to be run in order.
+          ret = cmd_exec(cmd, nil, datastore['Timeout'] || 15, { 'Channelized' => true })
+          if !ret
+            aborted = true
+          else
+            ret.strip!
+            aborted = true if !ret.empty? && ret !~ /The process tried to write to a nonexistent pipe./
+          end
+          if aborted
+            print_error('Error: Unable to execute the following command: ' + cmd.inspect)
+            print_error('Output: ' + ret.inspect) if ret && !ret.empty?
+            break
+          end
 
-        progress(total_bytes, sent)
+          sent += cmd.length
+
+          progress(total_bytes, sent)
+        end
       end
+      # The last command needs to be non-channelized, as otherwise that causes issues where the original session waits
+      # for an unlimited amount of time for the newly spawned session to exit.
+      cmd_exec(cmds.last, nil, datastore['Timeout'] || 15, { 'Channelized' => false })
```

Comment:
I wonder if we could revert this chunk of code to the previous functionality, and just toggle this flag individually? 🤔 

```ruby
cmds.each.with_index do |cmd, i|

  # ... as before

  # The last command should be fire-and-forget, otherwise issues occur where the original session waits
  # for an unlimited amount of time for the newly spawned session to exit.
  wait_for_cmd_result = i + 1 < cmds.length
  cmd_exec(cmds.last, nil, datastore['Timeout'] || 15, { 'Channelized' => wait_for_cmd_result })
```